pyensign/ensign.py

```diff
@@ -474,3 +489,203 @@ def _resolve_topic(self, topic):
             raise UnknownTopicError(
                 f"unknown topic '{topic}', please provide the name or ID of a topic in your project"
             )
+
+
+##########################################################################
+## Decorators
+##########################################################################
+
+# Global Ensign client. This is not thread-safe but it's safe to use concurrently in
+# asyncio coroutines.
+_client = None
+
+
+def authenticate(*auth_args, **auth_kwargs):
+    """
+    Decorator function to authenticate with Ensign. This function ideally should only
+    be called once, usually at the entry point in the application. Duplicate calls to
+    the decorator are ignored.
+
+    By default credentials are loaded from the ENSIGN_CLIENT_ID and ENSIGN_CLIENT_SECRET
+    environment variables. Alternatively, credentials can be provided as keyword
+    arguments or a JSON file path, see the Ensign class for more details.
+
+    Usage
+    -----
+    ```
+    # Use credentials from the environment
+    @authenticate()
+    async def main():
+        ...
+
+    # Use credentials from kwargs
+    @authenticate(client_id="...", client_secret="...")
+    async def main():
+        ...
+
+    # Use credentials from a JSON file
+    @authenticate(cred_path="...")
+    async def main():
+        ...
+    ```
+    """
+
+    def wrap_coroutine(coro):
+        async def wrapper(*args, **kwargs):
+            global _client
+            if _client is not None:
+                return await coro(*args, **kwargs)
```

Comment:
This prevents the client from being created twice, so a second call to `authenticate` will be a no-op. Although maybe this should actually be an error?